Allow multiple strings for missing values and correctly treat na = ""

andres-s opened the pull request:
Addresses #125 and #114.

I tried to follow the conventions of `utils::read.table`, namely `na = NULL` indicates no missing values while `na = ""` indicates that empty strings should be interpreted as missing values:

```r
example = "a,b
1,foo
,
"

f = textConnection(example)
read.csv(f, colClasses = c(a = "integer", b = "character"), na.strings = NULL)
#>    a   b
#> 1  1 foo
#> 2 NA    
close(f)

f = textConnection(example)
read.csv(f, colClasses = c(a = "integer", b = "character"), na.strings = "")
#>    a    b
#> 1  1  foo
#> 2 NA <NA>
close(f)
```

The implementation of `Token::flagNA` is a simple loop, I'm happy to do a performance comparison with less naive implementations if the behaviour is correct (although I suspect that the simple loop will win for small numbers of `na` strings).

hadley commented on R/tokenizer.R:
```diff
@@ -34,7 +34,8 @@ NULL
 #' @export
 #' @rdname Tokenizers
-#' @param na String to use for missing values.
+#' @param na Character vector of strings to use for missing values. Set this
```
I think it would be slightly simpler to use `character()` (i.e. a zero length character vector)` to indicate no missings

And maybe the default value for na should become `c("", "NA")`?

andres-s replied:
Happy to make both changes. `utils::read.csv` has `na.strings = "NA"` by default:
```r
example = "a,b
1,foo
,
"
f = textConnection(example)
read.csv(f, colClasses=c(a="integer", b="character"))
#>    a   b
#> 1  1 foo
#> 2 NA    
close(f)
```
As you can see though, it turns empty values into `NA` where the column type is logical, integer or numeric, so I think extending the same behaviour to character columns by default is pretty reasonable.

hadley replied:
Yeah, that makes sense to me too.

hadley commented:
This looks great! Apart from the changes mentioned in line, could you please also add a bullet point to `NEWS`, and look into why travis is failing?

andres-s commented:
Travis *was* failing because the [package version was insufficient](https://travis-ci.org/hadley/readr/builds/69886769#L1752). I cheekily amended the last commit to also bump the version to 0.1.1.9000 and the build is now passing... shall I drop the version back down to 0.1.0.9000 or keep it up? Note that at least one other [open pull request](https://github.com/hadley/readr/pull/183) is failing to build because of the same issue: https://travis-ci.org/hadley/readr/builds/67054495#L1666.

hadley commented:
Ah, that was because of the CRAN brouhaha. I've bumped in the master branch.

andres-s commented:
OK, I've updated `NEWS`, the default `na` option, changed `NULL` to `character()` and rebased. Let me know if there are any other adjustments to be made and I'll polish them off tomorrow.

hadley commented on NEWS.md:
```diff
@@ -17,3 +17,5 @@
 * When guessing field types, and there's no information to go on, use
   character instead of logical (#124, #128).
 
+* Multiple strings can now be specified to indicate missing values (#125).
```
Maybe also mention the default has changed?

hadley commented on src/Token.h:
```diff
@@ -82,14 +82,19 @@ class Token {
+      if ((size_t) (end_ - begin_) != (*it).size())
+        continue;
+
+      if (strncmp(begin_, &(*it)[0], (*it).size()) == 0) {
```
I think `it->size()` would be a bit simpler, and I assume you can simplify the other expression similarly. (`it->data()` maybe?)

andres-s replied:
Huh, I thought I would get a type error because `it->data()` would return a `std::string` pointer, but `strncmp` would expect a `char` pointer, but it works. I guess `std::string` is secretly `char[]` or something.

hadley commented on src/Tokenizer.cpp:
```diff
@@ -14,7 +14,14 @@ TokenizerPtr Tokenizer::create(List spec) {
-    std::string na = as<std::string>(spec["na"]);
+
+    std::vector<std::string> na;
+    if (Rf_isNull(spec["na"])) {
```
I think you can drop this check now

andres-s commented:
Thanks for picking up those points, I've made the changes and the code is much cleaner now.
